### ISSUE TITLE
[HTTP/3] New case of "The server returned an invalid or unrecognized response"

### ISSUE THREAD

**Issue by CarnaViire** [dotnet/runtime#57647]
Started happening after updating to `6.0.100-rc.1.21411.28`
First noticed in stress tests here https://github.com/dotnet/runtime/pull/57356#issuecomment-898642464
But it is not specific to stress, neither to status code, so it is not the same as #55988

I am able to (rarely) reproduce it by serially calling GET requests (harder to repro with Kestrel's logging enabled)

HttpClient error looks like this
```
System.Net.Http.HttpRequestException : The server returned an invalid or unrecognized response.
      Stack Trace:
        C:\Users\knatalia\dev\git\runtime\src\libraries\System.Net.Http\src\System\Net\Http\SocketsHttpHandler\Http3RequestStream.cs(318,0): at System.Net.Http.Http3RequestStream.ReadResponseAsync(CancellationToken cancellationToken)
        C:\Users\knatalia\dev\git\runtime\src\libraries\System.Net.Http\src\System\Net\Http\SocketsHttpHandler\Http3RequestStream.cs(196,0): at System.Net.Http.Http3RequestStream.SendAsync(CancellationToken cancellationToken)
        C:\Users\knatalia\dev\git\runtime\src\libraries\System.Net.Http\src\System\Net\Http\SocketsHttpHandler\Http3RequestStream.cs(297,0): at System.Net.Http.Http3RequestStream.SendAsync(CancellationToken cancellationToken)
        C:\Users\knatalia\dev\git\runtime\src\libraries\System.Net.Http\src\System\Net\Http\SocketsHttpHandler\Http3Connection.cs(210,0): at System.Net.Http.Http3Connection.SendAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
```

and related Kestrel log is
```
dbug: Microsoft.AspNetCore.Server.Kestrel[25]
      Connection id "0HMB29L9PVOIV", Request id "0HMB29L9PVOIV:00000058": started reading request body.
dbug: Microsoft.AspNetCore.Server.Kestrel[26]
      Connection id "0HMB29L9PVOIV", Request id "0HMB29L9PVOIV:00000058": done reading request body.
info: Microsoft.AspNetCore.Server.Kestrel[32]
      Connection id "0HMB29L9PVOIV", Request id "0HMB29L9PVOIV:00000058": the application completed without reading the entire request body.
dbug: Microsoft.AspNetCore.Server.Kestrel.Transport.Quic[13]
      Stream id "0HMB29L9PVOIV:00000058" read side aborted by application with error code 256 because: "The application completed without reading the entire request body.".
dbug: Microsoft.AspNetCore.Server.Kestrel.Transport.Quic[10]
      Stream id "0HMB29L9PVOIV:00000058" shutting down writes because: "The QUIC transport's send loop completed gracefully.".
dbug: Microsoft.AspNetCore.Server.Kestrel.Transport.Quic[2]
      Stream id "0HMB29L9PVOIV:0000005C" type Bidirectional accepted.
dbug: Microsoft.AspNetCore.Server.Kestrel.Transport.Quic[10]
      Stream id "0HMB29L9PVOIV:0000005C" shutting down writes because: "The QUIC transport's send loop completed gracefully.".
```

I think the problem might be tied to "the application completed without reading the entire request body".
Kestrel *did not call my callback* for that request. All the times the error happened without Kestrel's logging enabled, I've also seen that my callback was not called but HttpClient received some response which it couldn't read.

For the record, Kestrel's callback is set up like this
```c#
var statuses = Enum.GetValues<HttpStatusCode>()
    .Where(s => s >= HttpStatusCode.OK) // exclude informational
    .Cast<int>().Distinct(); // remove dupes like Ambiguous(300) and MultipleChoices(300) to avoid ambiguous mapping
foreach (var status in statuses)
{    
    app.MapGet("/Expect" + status, () => {
        Console.WriteLine("GET request to /Expect" + status + ", will return " + status + " " + (HttpStatusCode)status);
        //return new StatusCodeResult(status);
        return Results.StatusCode(status);
    });
}
```
and HttpClient's requests are
```c#
            using HttpClient client = CreateHttpClient();

            var statuses = Enum.GetValues<HttpStatusCode>()
                .Where(s => s >= HttpStatusCode.OK) // exclude informational
                .Cast<int>().Distinct(); // remove dupes like Ambiguous(300) and MultipleChoices(300) to avoid ambiguous mapping
            foreach (var status in statuses)
            {    
                Console.WriteLine("Request expecting " + status + " " + (HttpStatusCode)status);
                using HttpRequestMessage request = new()
                {
                    Method = HttpMethod.Get,
                    RequestUri = new Uri("https://localhost:5001/Expect" + status),
                    Version = HttpVersion30,
                    VersionPolicy = HttpVersionPolicy.RequestVersionExact
                };
                using HttpResponseMessage response = await client.SendAsync(request).WaitAsync(TimeSpan.FromSeconds(10));
                
                if(status != (int)response.StatusCode)
                {
                    fail = true;
                    Console.WriteLine("    FAILURE -- received " + (int)response.StatusCode + " " + response.StatusCode);
                }
                else
                {
                    Console.WriteLine("    SUCCESS");
                }
            }
```

cc @JamesNK

**Comment by CarnaViire** [comment]
@JamesNK in what situation can "the application completed without reading the entire request body" happen? What does this mean and why the callback was not called?

**Comment by ManickaP** [comment]
@CarnaViire could you in your test catch the exception from the client and continue sending additional requests? I'd like to confirm/deny my theory with another type of errors seen in the stress.

**Comment by JamesNK** [comment]
I think this is an issue in Kestrel and will be fixed with https://github.com/dotnet/aspnetcore/pull/35434

> @JamesNK in what situation can "the application completed without reading the entire request body" happen? What does this mean and why the callback was not called?

It happens if the request body isn't read until the end. The logs indicate you have started reading it, but not until the end.

**Comment by CarnaViire** [comment]
@ManickaP you are right. On the next request after `The server returned an invalid or unrecognized response.` I got `QuicConnectionAbortedException: Connection aborted by peer (0).` The following requests were fine.

**Comment by ManickaP** [comment]
Thanks for confirmation @CarnaViire, I filed another issue in ASP https://github.com/dotnet/aspnetcore/issues/35470.

**Comment by CarnaViire** [comment]
To sum up, we expect this issue to get resolved after dotnet/aspnetcore#35434, dotnet/aspnetcore#35470 and https://github.com/dotnet/aspnetcore/issues/35399 get fixed/merged.

**Comment by karelz** [comment]
@CarnaViire both issues are now merged. I guess we can close this issue and unblock our Stress.
cc @ManickaP

**Comment by CarnaViire** [comment]
Today's stress run still has the issue (`System.Net.Http.HttpRequestException: The server returned an invalid or unrecognized response.`) https://dev.azure.com/dnceng/public/_build/results?buildId=1311648&view=logs&j=0da5d1d9-276d-5173-c4c4-9d4d4ed14fdb&t=1451f5f3-0108-5a08-5b92-e984b2a85bbd
ASP.NET Core version is `6.0.0-rc.1.21419.28`
@JamesNK do you know if this version includes both fixes?

**Comment by JamesNK** [comment]
I don't think so.

It's 4 days old - https://github.com/dotnet/sdk/pull/19917
These changes were merged into 6.0-rc1 3 days ago - https://github.com/dotnet/aspnetcore/commits/release/6.0-rc1

**Comment by JamesNK** [comment]
The stress runs on that environment have used an older SDK over the last couple of days. You should investigate why and get it to use the latest SDK.

**Comment by ManickaP** [comment]
We're running in the nightly docker image and that hasn't been updated yet, i.e.: https://github.com/dotnet/dotnet-docker/blob/nightly/src/runtime/6.0/bullseye-slim/amd64/Dockerfile

**Comment by ManickaP** [comment]
This particular issue is fixed, H/3 stress runs for 30 mins, no occurrences of this error, closing.
Other errors do happen, but we track that elsewhere.

**Comment by karelz** [comment]
Addressed in 6.0, reflecting that in the milestone.
